### PR TITLE
op-node: EIP-4844 support

### PR DESCRIPTION
**Pull request by roberto-bayardo**
**Description**

Original rebased prototype by proto, plus changes from Roberto:
- encapsulate data <-> blob conversion code, add unit tests
- update 4844 code to be compatible with latest beacon node api
- remove stray file, include one more invalid blob test
- appropriate fee bumping & blob fee estimation for blob transactions
- misc other improvements

**Tests**

Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.

**Additional context**

https://github.com/orgs/ethereum-optimism/projects/47/views/2?pane=issue&itemId=37047038

### CODE REVIEW

File: op-batcher/batcher/driver.go [hunk 1]
```diff
@@ -334,22 +334,37 @@ func (l *BatchSubmitter) publishTxToL1(ctx context.Context, queue *txmgr.Queue[t
...
 
-// sendTransaction creates & submits a transaction to the batch inbox address with the given `data`.
+// sendTransaction creates & submits a transaction to the batch inbox address with the given `txData`.
 // It currently uses the underlying `txmgr` to handle transaction sending & price management.
 // This is a blocking method. It should not be called concurrently.
 func (l *BatchSubmitter) sendTransaction(txdata txData, queue *txmgr.Queue[txData], receiptsCh chan txmgr.TxReceipt[txData]) {
 	// Do the gas estimation offline. A value of 0 will cause the [txmgr] to estimate the gas limit.
 	data := txdata.Bytes()
+
+	var blobs []*eth.Blob
```
Comment (tynes) [inline op-batcher/batcher/driver.go:363]: Why is this safe to operate on a local timestamp unlike other hardfork logic that operates on a L1 based timestamp?

Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go:363]: Good point. This triggers whether to send a blob instead of regular transaction for the batch. If the L1 state is actually behind our local time then we could end up with a blob transaction included in a pre-Blob activation block that would get ignored by chain derivation leading to l2 reorg.

We could explicitly check to make sure chain tip is beyond BlobsEnabledL1Timestamp, though I guess to be super certain we won't have an L2 re-org because of this, we should wait some # of confirmations after?

Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go:344]: Note:  In a subsequent WIP PR I'm introducing another guard on whether the batcher will use blob transactions based on the SysConfig state (https://github.com/roberto-bayardo/optimism/pull/2).  I think this change will resolve this issue once I implement the TODO there to only activate blobs if the L1 block number from the sysconfig update event is beyond the blob activation time.

Comment (tynes) [reply op-batcher/batcher/driver.go:363]: Hmm I don't think that we have risk of a L2 reorg without waiting a # of confirmations after - the invariant must hold that `BlobsEnabledL1Timestamp >= L1 DecunTimestamp`. Could you explain in more depth the scenario you are thinking of?

We also want to not have this solely dependent on the timestamp, we want to have an override that can turn it on/off because for chains with very low throughput it will likely make it more expensive

File: op-chain-ops/genesis/config.go [hunk 2]
```diff
@@ -564,6 +578,8 @@ func (d *DeployConfig) RollupConfig(l1StartBlock *types.Block, l2GenesisBlockHas
 		EclipseTime:            d.EclipseTime(l1StartBlock.Time()),
 		FjordTime:              d.FjordTime(l1StartBlock.Time()),
 		InteropTime:            d.InteropTime(l1StartBlock.Time()),
+		// 4844 blobs usage activation for rollup DA
+		BlobsEnabledL1Timestamp: d.BlobsUpgradeTime(l1StartBlock.Time()),
```
Comment (tynes) [inline op-chain-ops/genesis/config.go]: I believe that we should remove blobs enabled timestamp and use eclipse timestamp instead but should double check with @protolambda

File: op-node/rollup/derive/blob_data_source.go [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+		if _, txs, err := ds.fetcher.InfoAndTxsByHash(ctx, ds.ref.Hash); err == nil {
+			ds.open = true
+			ds.callData, ds.blobDataHashes = BlobDataFromEVMTransactions(ds.cfg, ds.batcherAddr, txs, ds.log)
+		} else if errors.Is(err, ethereum.NotFound) {
+			return nil, NewResetError(fmt.Errorf("failed to open blob-data source: %w", err))
+		} else {
+			return nil, NewTemporaryError(fmt.Errorf("failed to open blob-data source: %w", err))
+		}
+	}
+	// prioritize call-data
```
Comment (tynes) [inline op-node/rollup/derive/blob_data_source.go:61]: This behavior needs to be very explicit in the specs and perhaps we want a sanity check in the batcher to make it very explicit across the codebase there that SHOULD only be calldata or blobs and blobs are ignored if calldata is present

It silently ignores the blobs here when calldata is present, if it doesn't add too much complexity to the implementation it would be nice to defensively log something.

I am not sure if we want to exist in a world where we can derive both calldata and blobs from the same transaction so i understand the approach here

File: op-chain-ops/genesis/config.go (hunk 2 above)
Comment (protolambda) [reply op-chain-ops/genesis/config.go]: It's fine to leave it as-is with `BlobsEnabledL1Timestamp` since it's two hardforks out (`canyon` and `delta` first), but yes, ideally we prepare this for the `eclipse` upgrade proposal after.

File: op-batcher/batcher/driver.go (hunk 1 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go:363]: Issue is the batcher can't control exactly which L1 block a batcher transaction appears in regardless of using local timestamp or L1 derived timestamp. Consider this: batcher might submit a blob tx to the L1 txpool after blobsenabledL1 timestamp, but then the L1 might reorg to a point before the blobsenabledL1 timestamp, which can result in the blob tx to appear in an L1 block before the blobsenabledtimestamp.  The result here is that chain derivation would ignore that blob transaction.

Maybe this is a non-issue once I introduce the config option to turn batcher's use of blobs on/off. We can default it to "off" and only recommend turning it on well after the blobsEnabledTimestamp has passed.

File: op-node/rollup/derive/blob_data_source.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:61]: Blob data would not be ignored: if there was blobdata from the same transaction it would be returned by a later call to Next().  See the implementation in BlobDataFromEVMTransactions below which is where both calldata and blob data hashes are extracted (possibly from a single tx). That said the batcher will not (initially) be using both calldata and blob data, but we'd have the flexibility to add that later if it turned out to be worth doing at some point.

Comment (tynes) [reply op-node/rollup/derive/blob_data_source.go:61]: Got it ok thanks! To clarify, it is not an exclusive or but an and where the calldata is consumed first and then the blob data?

The espresso leader election work introduced the idea of putting a "batch header" in the calldata and turning the batch inbox into a smart contract and altering the batch auth logic to be correct tx.to and non reverting receipt instead of correct tx.to and correct tx.from. If we want to be forwards compatible with that design, we probably want to ignore calldata if a blob is present then in the future we would check that a valid header exists in the calldata if the leader election is turned on. The batch header would just include metadata that would allow for a smart contract to revert if the sender isn't the current leader

File: op-e2e/actions/l2_verifier.go [hunk 4]
```diff
@@ -59,7 +59,8 @@ type L2API interface {
 
 func NewL2Verifier(t Testing, log log.Logger, l1 derive.L1Fetcher, eng L2API, cfg *rollup.Config, syncCfg *sync.Config) *L2Verifier {
 	metrics := &testutils.TestDerivationMetrics{}
-	pipeline := derive.NewDerivationPipeline(log, cfg, l1, eng, metrics, syncCfg)
+	// TODO blob testing
```
Comment (tynes) [inline op-e2e/actions/l2_verifier.go:62]: We shouldn't merge with this open todo or we should link an issue

File: op-node/node/node.go [hunk 5]
```diff
@@ -291,6 +296,22 @@ func (n *OpNode) initRuntimeConfig(ctx context.Context, cfg *Config) error {
...
+func (n *OpNode) initL1BeaconAPI(ctx context.Context, cfg *Config) error {
+	if cfg.Beacon == nil {
+		return nil
+	}
+	httpClient, err := cfg.Beacon.Setup(ctx, n.log)
+	if err != nil {
+		return fmt.Errorf("failed to setup L2 execution-engine RPC client: %w", err)
+	}
+
+	// TODO: wrap http client with metrics
```
Comment (tynes) [inline op-node/node/node.go:305]: This todo should have a linked issue

File: op-node/rollup/derive/blob_data_source.go (hunk 3 above)
Comment (tynes) [reply op-node/rollup/derive/blob_data_source.go:61]: I would like the exact behavior here documented in the specs so that we can match the specified behavior against the implementation before merging this to ensure there are no ambiguities

File: op-node/rollup/derive/blob_data_source.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+// from transactions that are sent to the batch inbox address from the batch sender address.
+// This will return an empty array if no valid transactions are found.
+// Call-data can be used as fallback in case blobs are overpriced or unstable.
+func BlobDataFromEVMTransactions(dsCfg DataSourceConfig, batcherAddr common.Address, txs types.Transactions, log log.Logger) ([]eth.Data, []eth.IndexedDataHash) {
+	var callData []eth.Data
+	var indexedDataHashes []eth.IndexedDataHash
+	blobIndex := uint64(0)
+	for j, tx := range txs {
+		if to := tx.To(); to != nil && *to == dsCfg.batchInboxAddress {
+			seqDataSubmitter, err := dsCfg.l1Signer.Sender(tx) // optimization: only derive sender if To is correct
```
Comment (tynes) [inline op-node/rollup/derive/blob_data_source.go]: I think we could dedupe the shared code between this function and the calldata based function in a small helper function

File: op-service/client/http.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+	}
+	for k, values := range headers {
+		for _, v := range values {
+			req.Header.Add(k, v)
+		}
+	}
+	return cl.client.Do(req)
+}
+
+// TODO HTTP client wrapper to track response time and error metrics
```
Comment (tynes) [inline op-service/client/http.go:39]: This todo should have a linked ticket

File: op-service/txmgr/txmgr.go [hunk 8]
```diff
@@ -12,23 +12,30 @@ import (
...
 
 const (
-	// Geth requires a minimum fee bump of 10% for tx resubmission
-	priceBump int64 = 10
+	// Geth requires a minimum fee bump of 10% for regular tx resubmission, 100% for blob txs
+	priceBump     int64 = 10
+	blobPriceBump int64 = 100
 )
 
 // new = old * (100 + priceBump) / 100
```
Comment (tynes) [inline op-service/txmgr/txmgr.go]: small nit

```golang
var (
   ...
)
```

**Comment by tynes** [comment]
This needs a rebase and also it should not alter the `forge-std` version

File: op-batcher/batcher/driver.go (hunk 1 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go:363]: I'm going to resolve this conversation based on the call we had today. We'll implement a runtime "Strategy" flag that's up for the chain operator to set appropriately (e.g. it would never be set on a chain where Eclipse wasn't already active).

File: op-node/rollup/derive/blob_data_source.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:61]: Based on call today we'll change this so that if you post as a batch using a blob TX then you cannot (also) post a batch in its calldata (so it's either or on a per TX level).

File: op-node/rollup/derive/blob_data_source.go (hunk 6 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:102]: it's subtly different though: the blob version has to worry about multiple blobs (and hence batches) per tx requiring a nested loop; the calldata version doesn't need to worry about this.   Blob version also returns pointers to the data 
 (which then need to be requested from the CL) vs. the actual data in the calldata version.

Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go]: Oh the tx validity checks can be deduped, looking at that.

File: op-service/txmgr/txmgr.go (hunk 8 above)
Comment (roberto-bayardo) [reply op-service/txmgr/txmgr.go]: fixed

File: op-node/rollup/derive/blob_data_source.go (hunk 6 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go]: pulled out the tx validity check into its own func for reuse

File: op-node/rollup/derive/blob_data_source.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:61]: Updated the derivation code so that calldata in blob transactions is ignored.

Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:61]: Spec change: https://github.com/ethereum-optimism/optimism/pull/8660

File: op-service/eth/blob.go [hunk 9]
```diff
@@ -66,3 +67,67 @@ func KZGToVersionedHash(commitment kzg4844.Commitment) (out common.Hash) {
 func VerifyBlobProof(blob *Blob, commitment kzg4844.Commitment, proof kzg4844.Proof) error {
 	return kzg4844.VerifyBlobProof(*blob.KZGBlob(), commitment, proof)
 }
+
+// FromData encodes the given input data into this blob. The encoding scheme is as follows:
+//
+// First, field elements are encoded as big-endian uint256 in BLS modulus range. To avoid modulus
+// overflow, we can't use the full 32 bytes, so we write data only to the topmost 31 bytes of each.
+// TODO: we can optimize this to get a bit more data from the blobs by using the top byte
```
Comment (trianglesphere) [inline op-service/eth/blob.go:75]: I think the specs say that we do eek out those last 3 bytes per 4 field elements

Comment (roberto-bayardo) [reply op-service/eth/blob.go:75]: Right, Proto added this to the spec yesterday and it's in the process of implementation in another PR.

File: op-batcher/batcher/driver.go [hunk 10]
```diff
@@ -353,22 +353,37 @@ func (l *BatchSubmitter) publishTxToL1(ctx context.Context, queue *txmgr.Queue[t
...
+	if l.RollupConfig.BlobsEnabledL1Timestamp != nil && *l.RollupConfig.BlobsEnabledL1Timestamp <= uint64(time.Now().Unix()) {
```
Comment (tynes) [inline op-batcher/batcher/driver.go:364]: Regarding the logic that selects between blobs or calldata:

- We should support a --da-type flag with two values: calldata, blobs. In the future we’ll add alt-da.
- Extract the DA submission into an interface to allow each DA type to be implemented in isolation and for more to be easily added.

The first would be great, the second would be nice to have and could be a follow up

File: op-node/rollup/superchain.go [hunk 11]
```diff
@@ -100,7 +100,7 @@ func LoadOPStackRollupConfig(chainID uint64) (*Config, error) {
 		RegolithTime:           &regolithTime,
 		CanyonTime:             superChain.Config.CanyonTime,
 		DeltaTime:              superChain.Config.DeltaTime,
-		EclipseTime:            superChain.Config.EclipseTime,
+		EcotoneTime:            superChain.Config.EclipseTime,
```
[the remark this answers is not in the document]
Comment (tynes) [reply op-node/rollup/superchain.go:103]: This is a good catch

File: op-service/eth/blob.go [hunk 12]
```diff
@@ -66,3 +67,67 @@ func KZGToVersionedHash(commitment kzg4844.Commitment) (out common.Hash) {
...
+// partially.
```
Comment (tynes) [inline op-service/eth/blob.go:76]: Perhaps we can break this out into its own PR

File: op-service/sources/l1_beacon_client.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+		_ = resp.Body.Close()
+		return err
+	}
+	if err := resp.Body.Close(); err != nil {
+		return fmt.Errorf("failed to close genesis response object: %w", err)
+	}
+	return nil
+}
+
+func (cl *L1BeaconClient) GetTimeToSlotFn(ctx context.Context) (TimeToSlotFn, error) {
```
Comment (tynes) [inline op-service/sources/l1_beacon_client.go:55]: I wouldn't be opposed to having the client broken out into its own PR and merged first

File: op-service/txmgr/txmgr.go [hunk 14]
```diff
@@ -554,61 +614,107 @@ func (m *SimpleTxManager) increaseGasPrice(ctx context.Context, tx *types.Transa
...
+	if head.ExcessBlobGas != nil {
+		blobFee = eip4844.CalcBlobFee(*head.ExcessBlobGas)
+	}
+	return tip, head.BaseFee, blobFee, nil
 }
 
 // calcThresholdValue returns x * priceBumpPercent / 100
-func calcThresholdValue(x *big.Int) *big.Int {
-	threshold := new(big.Int).Mul(priceBumpPercent, x)
+func calcThresholdValue(x *big.Int, isBlobTx bool) *big.Int {
```
Comment (tynes) [inline op-service/txmgr/txmgr.go:694]: How do you feel about two functions that wrap this that hardcode the bool to make the caller api dead simple?

```go
func calcThresholdValueBlobTx(x *big.Int) *big.Int {
  return calcThresholdValue(x, true)
}

func calcThresholdValueDynamicTx(x *big.Int) *big.Int {
  return calcThresholdValue(x, false)
}

```

**Comment by tynes** [comment]
It would be nice if this PR could be broken up into smaller chunks to facilitate easier review.

It also needs the system config changes for the new enum that can be emitted in the `ConfigUpdate` event

**Comment by roberto-bayardo** [comment]
Closing this to break into smaller PRs as suggested by @tynes
